lib/pf/mfa/Akamai.pm

```diff
@@ -0,0 +1,507 @@
+        if ($otp =~ /^\d{6,6}$/ || $otp =~ /^\d{16,16}$/) {
+            if ( grep $_ eq 'totp', @{$default_device[0]->{'methods'}}) {
+                return $ACTIONS{'totp'}->($self,$default_device[0]->{'device'},$username,$otp);
+            }
+        } elsif ($otp =~ /^\d{8,8}$/) {
+                return $ACTIONS{'check_auth'}->($self,$default_device[0]->{'device'},$username,$otp);
+        } elsif ($otp =~ /^(sms|push|phone)(\d?)$/i) {
+            my @device = $self->select_phone($devices->{'result'}->{'devices'}, $2);
+            my $method = $1;
+            foreach my $device (@device) {
+                if ( grep $_ =~ $METHOD_ALIAS{$method}, @{$device->{'methods'}}) {
+                    return $ACTIONS{$method}->($self,$device->{'device'},$username,$1);
+                } else {
+                    $logger->warn("Unsuported method on device ".$device->{'name'});
+                    return $FALSE;
+                }
+            }
+        } else {
+            $logger->warn("Method not supported");
+            return $FALSE;
+        }
+    }
+}
+
+=head2 select_phone
+
+Select the phone to trigger the MFA
+
+=cut
+
+sub select_phone {
+    my ($self, $devices, $phone_id) = @_;
+    my $logger = get_logger();
+    my @device;
+    if (defined($phone_id) && $phone_id ne "") {
+        if ($phone_id == 1 || $phone_id == 0) {
+            # Return the default phone
+            @device = grep { $_->{'default'} == 1} @{$devices};
+        }
+        # Return the n-1 phone
+        @device = @{$devices}[$phone_id-1];
+    } else {
+        # Return the default phone
+        @device = grep { $_->{'default'} == 1 } @{$devices};
+    }
+    return @device;
+}
+
+=head2 totp
+
+totp method
+
+=cut
+
+sub totp {
+    my ($self, $device, $username, $otp) = @_;
+    my $logger = get_logger();
+    my $method = "offline_otp";
+    if (length($otp) == 16) {
+        $method = "bypass_code";
+    }
+    $logger->info("Trigger $method for user $username");
+    my $post_fields = encode_json({device => $device, method => { $method => {"code" => $otp} } , username => $username});
+    my ($auth, $error) = $self->_post_curl("/api/v1/verify/start_auth", $post_fields);
+    if ($error) {
+        return $FALSE;
+    }
+    if ($auth->{'result'}->{'status'} eq 'allow') {
+        $logger->warn("Authentication sucessfull on Akamai MFA");
+        return $TRUE;
+    }
+    $logger->warn("Authentication denied on Akamai MFA, reason: ". $auth->{'result'}->{'status'}->{'deny'}->{'reason'});
+    return $FALSE;
+}
+
+=head2 generic_method
+
+generic method
+
+=cut
+
+sub generic_method {
+    my ($self, $device, $username, $method) =@_;
+    my $logger = get_logger();
+    $logger->info("Trigger $method for user $username");
+    my $post_fields = encode_json({device => $device, method => $METHOD_LOOKUP{$method}, username => $username});
+    my $chi = pf::CHI->new(namespace => 'mfa');
```

Comment:
make this a `sub cache { return pf::CHI->new(namespace => 'mfa'); }` and use it everywhere